File: app/src/main/java/com/robifr/ledger/ui/editcustomer/EditCustomerFragment.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.editcustomer;

import android.os.Bundle;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavBackStackEntry;
import androidx.navigation.Navigation;
import com.robifr.ledger.R;
import com.robifr.ledger.data.model.CustomerModel;
import com.robifr.ledger.ui.FragmentResultKey;
import com.robifr.ledger.ui.createcustomer.CreateCustomerFragment;
import com.robifr.ledger.ui.editcustomer.viewmodel.EditCustomerViewModel;
import java.util.Objects;

public class EditCustomerFragment extends CreateCustomerFragment {
  public enum Arguments implements FragmentResultKey {
    INITIAL_CUSTOMER_ID_TO_EDIT;

    @Override
    @NonNull
    public String key() {
      return FragmentResultKey.generateKey(this);
    }
  }

  public enum Request implements FragmentResultKey {
    EDIT_CUSTOMER;

    @Override
    @NonNull
    public String key() {
      return FragmentResultKey.generateKey(this);
    }
  }

  public enum Result implements FragmentResultKey {
    EDITED_CUSTOMER_ID;

    @Override
    @NonNull
    public String key() {
      return FragmentResultKey.generateKey(this);
    }
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstance) {
    super.onViewCreated(view, savedInstance);
    Objects.requireNonNull(this._fragmentBinding);

    this._createCustomerViewModel =
        new ViewModelProvider(this, new EditCustomerViewModel.Factory(this.requireContext()))
            .get(EditCustomerViewModel.class);
    this._viewModelHandler =
        new EditCustomerViewModelHandler(
            this, (EditCustomerViewModel) this._createCustomerViewModel);

    this._fragmentBinding.toolbar.setTitle(this.getString(R.string.text_edit_customer));

    final NavBackStackEntry backStackEntry =
        Navigation.findNavController(this._fragmentBinding.getRoot()).getCurrentBackStackEntry();

    if (this._createCustomerViewModel instanceof EditCustomerViewModel editCustomerViewModel
        && backStackEntry != null
        && backStackEntry.getArguments() != null) {
      final CustomerModel initialCustomer =
          editCustomerViewModel.selectCustomerById(
              backStackEntry
                  .getArguments()
                  .getLong(Arguments.INITIAL_CUSTOMER_ID_TO_EDIT.key(), 0L));
      Objects.requireNonNull(initialCustomer); // Logically shouldn't be null when editing data.

      editCustomerViewModel.setInitialCustomerToEdit(initialCustomer);
      editCustomerViewModel.onNameTextChanged(initialCustomer.name());
      editCustomerViewModel.onBalanceChanged(initialCustomer.balance());
      editCustomerViewModel.onDebtChanged(initialCustomer.debt());
    }
  }
}

File: app/src/main/java/com/robifr/ledger/ui/editcustomer/EditCustomerFragment.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.editcustomer

import android.os.Bundle
import android.view.View
import androidx.fragment.app.viewModels
import com.robifr.ledger.R
import com.robifr.ledger.ui.FragmentResultKey
import com.robifr.ledger.ui.createcustomer.CreateCustomerFragment
import com.robifr.ledger.ui.editcustomer.viewmodel.EditCustomerResultState
import com.robifr.ledger.ui.editcustomer.viewmodel.EditCustomerViewModel
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class EditCustomerFragment : CreateCustomerFragment() {
  override val createCustomerViewModel: EditCustomerViewModel by viewModels()

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    fragmentBinding.toolbar.setTitle(R.string.editCustomer)
    createCustomerViewModel.editResultState.observe(viewLifecycleOwner, ::_onResultState)
  }

  private fun _onResultState(state: EditCustomerResultState) {
    parentFragmentManager.setFragmentResult(
        Request.EDIT_CUSTOMER.key,
        Bundle().apply {
          state.editedCustomerId?.let { putLong(Result.EDITED_CUSTOMER_ID_LONG.key, it) }
        })
    finish()
  }

  enum class Arguments : FragmentResultKey {
    INITIAL_CUSTOMER_ID_TO_EDIT_LONG
  }

  enum class Request : FragmentResultKey {
    EDIT_CUSTOMER
  }

  enum class Result : FragmentResultKey {
    EDITED_CUSTOMER_ID_LONG
  }
}
